### ISSUE TITLE
expo run:android fails when adding productFlavors with custom applicationId

### ISSUE THREAD

**Issue by fikkatra** [expo/expo#18315]
### Summary

When using the Bare workflow and adding productFlavors for Android, an error occurs after building and installing the app locally on an emulator. This error only occurs when an applicationId is added to the productFlavors. When removing the applicationId from the productFlavor, everyting works fine and the other variables within the flavors get picked up nicely.

### Managed or bare workflow? If you have made manual changes inside of the `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### Package versions

"expo": "~45.0.0",
"expo-splash-screen": "~0.15.1",
"expo-status-bar": "~1.3.0",
"react": "17.0.2",
"react-dom": "17.0.2",
"react-native": "0.68.2",
"react-native-web": "0.17.7"

### Environment

expo-env-info 1.0.4 environment info:
    System:
      OS: macOS 12.4
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 8.11.0 - ~/.nvm/versions/node/v16.16.0/bin/npm
      Watchman: 2022.07.04.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
      Android SDK:
        API Levels: 30, 31
        Build Tools: 29.0.2, 30.0.2, 30.0.3
        System Images: android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 4.2 AI-202.7660.26.42.7486908
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmGlobalPackages:
      expo-cli: 5.5.1

### Reproducible demo

Repo can be found [here](https://github.com/fikkatra/expo-variant-error-repro).

Also reproducible with the following steps:

Fresh install of a new app with the Bare workflow:

> npx create-expo-app --template bare-minimum

Update android/app/guild.gradle:

Add productFlavors:
```
  flavorDimensions "env"
  productFlavors {
      production {
          dimension "env"
          applicationId 'com.myapp'
      }
      development {
          dimension "env"
          applicationId 'com.myapp.dev'
      }
  }
```

Add the following line under buildTypes -> debug:
`matchingFallbacks = ['debug', 'release']`

Run:
> expo run:android --variant=developmentdebug

### Stacktrace (if a crash is involved)

› Building app...
Configuration on demand is an incubating feature.

> Configure project :app
Unable to detect AGP versions for included builds. All projects in the build should use the same AGP version. Class name for the included build object: org.gradle.composite.internal.DefaultIncludedBuild$IncludedBuildImpl_Decorated.

> Configure project :expo

Using expo modules
  - expo-application (4.1.0)
  - expo-constants (13.1.1)
  - expo-error-recovery (3.1.0)
  - expo-file-system (14.0.0)
  - expo-font (10.1.0)
  - expo-keep-awake (10.1.1)
  - expo-modules-core (0.9.2)
  - expo-splash-screen (0.15.1)

> Task :expo-constants:createDebugExpoConfig
Execution optimizations have been disabled for task ':expo-constants:createDebugExpoConfig' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:checkDevelopmentDebugAarMetadata' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:compileDevelopmentDebugAidl' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:compileDevelopmentDebugRenderscript' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:compileDevelopmentDebugShaders' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:generateDevelopmentDebugBuildConfig' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:javaPreCompileDevelopmentDebug' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:mergeDevelopmentDebugShaders' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD SUCCESSFUL in 9s
211 actionable tasks: 5 executed, 206 up-to-date
Starting Metro Bundler
› Metro waiting on http://192.168.0.118:8081
› Linking is disabled because the client scheme cannot be resolved.

› Press a │ open Android
› Press i │ open iOS simulator
› Press w │ open web

› Press r │ reload app
› Press m │ toggle menu
› Press d │ show developer tools
› shift+d │ toggle auto opening developer tools on startup (disabled)

› Press ? │ show all commands

› Failed to locate binary file, installing with Gradle...
Started Metro Bundler

> Configure project :app
Unable to detect AGP versions for included builds. All projects in the build should use the same AGP version. Class name for the included build object: org.gradle.composite.internal.DefaultIncludedBuild$IncludedBuildImpl_Decorated.

> Configure project :expo

Using expo modules
  - expo-application (4.1.0)
  - expo-constants (13.1.1)
  - expo-error-recovery (3.1.0)
  - expo-file-system (14.0.0)
  - expo-font (10.1.0)
  - expo-keep-awake (10.1.1)
  - expo-modules-core (0.9.2)
  - expo-splash-screen (0.15.1)

> Task :expo-constants:createDebugExpoConfig
Execution optimizations have been disabled for task ':expo-constants:createDebugExpoConfig' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:checkDevelopmentDebugAarMetadata' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:compileDevelopmentDebugAidl' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:compileDevelopmentDebugRenderscript' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:compileDevelopmentDebugShaders' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:generateDevelopmentDebugBuildConfig' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:javaPreCompileDevelopmentDebug' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/x/Projects/my-app/android'. Reason: Task ':expo-constants:createDebugExpoConfig' uses this output of task ':app:mergeDevelopmentDebugShaders' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.

> Task :app:installDevelopmentDebug
Installing APK 'app-development-debug.apk' on 'Pixel_4_API_30(AVD) - 11' for app:development-debug
Installed on 1 device.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD SUCCESSFUL in 11s
212 actionable tasks: 6 executed, 206 up-to-date
Couldn't open Android app with activity "com.myapp/.MainActivity" on device "Pixel_4_API_30".
The app might not be installed, try installing it with: expo run:android -d Pixel_4_API_30
Activity class {com.myapp/com.myapp.MainActivity} does not exist.

**Comment by brentvatne** [comment]
@fikkatra - can you try reproducing this on SDK 46 beta with the local CLI?

```
EXPO_BETA=1 npx create-expo-app --template bare-minimum
```

and then to invoke run  android, within the project do:

```
npx expo run:android
```

**Comment by fikkatra** [comment]
@brentvatne thanks for your reply. When scaffolding with SDK 46 beta and running `npx expo run:android`, I get the following error:
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Task 'installDebug' is ambiguous in project ':app'. Candidates are: 'installDevelopmentDebug', 'installDevelopmentDebugAndroidTest', 'installProductionDebug', 'installProductionDebugAndroidTest'.
> 

This makes sense because I didn't pass a variant, but running `npx expo run:android --variant=developmentdebug` throws the following error:

> CommandError: Unknown arguments: --variant=developmentdebug

When running `expo run:android --variant=developmentdebug` (without npx), I get the same error as before.

Some more exploration: I tried running the app with `react-native run-android --variant=developmentdebug` and got the same error. It seems to be the same issue as [this one](https://stackoverflow.com/questions/39289591/react-native-productflavors-error-type-3-activity-class-does-not-exist). Indeed, running `react-native run-android --variant=developmentdebug --appId com.myapp.dev` works. However, expo CLI does not seem to support this. When running `expo run:android --variant=developmentdebug --appId com.exporepro.dev` I get the error:

> CommandError: Unknown arguments: --variant=developmentdebug, --appId

**Comment by fonseca80lines** [comment]
I faced the same problem. This occurs in at least 2 different issues.
Analyzing the "expo-cli/commands/run/android/androidRun.ts" file from the expo-cli package I saw that the first problem is in the resolveOptionsAsync method in assigning the const buildType and flavors with the variant parameter.
![image](https://user-images.githubusercontent.com/95696597/186242697-fb11a4a7-63b7-40c0-b4ce-ab786b4aa0e0.png)
The first argument should be the flavor and the second the buildType like "developmentDebug", Arguments other than this cause an error in the build for not finding the task.
This way apkVariantDirectory gets inverted folder like "android/app/build/outputs/apk/debug/development" while it should be "android/app/build/outputs/apk/development/debug" causing the first error "Failed to locate binary file, installing with Gradle..."


The second problem is in the file "config-plugins/android/Package.ts" in the method "getApplicationIdAsync" which looks for the first match with "applicationId "com.app"" that it finds in the build.gradle file and assumes it as appId. In this case, if there is applicationId in defaultConfigs of build.gradle, it will assume it as appId, or else it will get the applicationId in the first flavor configured with this parameter.
![image](https://user-images.githubusercontent.com/95696597/186242887-15404d2b-1506-44a1-b3c6-b3ad4145d2cd.png)

**Comment by bentburton** [comment]
Facing same issue. any update?

**Comment by jlmart88** [comment]
I am also facing this issue with Expo 46

**Comment by alfjesus** [comment]
Same problem for me with Expo 47

**Comment by hernan-clicoh** [comment]
Works when using the global expo cli not working with npx run:android, or when using a script on package.json

**Comment by anuragbansals** [comment]
Facing same issue. I want to use different flavors as per my different client names very different from development and production. When I use the command expo run:android --variant=variant_name --device, it is downloading the app in my device but not starting the local server. It throws the error `No development build (app_id) for this project is installed. Please make and install a development build on the device first.`

**Comment by alfjesus** [comment]
@anuragbansals About a year ago I was struggling with the same on several occasions. I finally made the choice to migrate to Expo CNG and Expo prebuild. Wish I had done the switch earlier. With a dynamic app config (https://docs.expo.dev/workflow/configuration/#dynamic-configuration) you can easily generate/run the flavor (android) you want (i.e dev, stage and prod) based on runtime variables.

I highly recommend for you to consider doing the same switch 🙂

**Comment by chakravarthi-bb** [comment]
I am still facing this issue, trying to integrate expo from bare react native project

**Comment by brentvatne** [comment]
@chakravarthi-bb - would you be able to create a new issue with an up-to-date minimal reproducible example that demonstrates this? we can definitely investigate if so
